### ISSUE TITLE
Dereferencing hdk::api lazy_static structs in validation causes WASM Trap Host(BadCallError)

### ISSUE THREAD

**Issue by MightyAlex200** [holochain/holochain-rust#1366]
`hc`, `hdk` version: `0.0.13-alpha1`.

In the `validation` section of the `hdk::entry!` macro, dereferencing any of the `lazy_static!` refs in `/hdk-rust/src/api.rs` will cause a `Host(BadCallError)` error.

Minimal case:
1. Use `hc init` and `hc generate zomes/my_zome rust` to create a simple working application.
1. Run `git apply` on the following diff file:
```diff
diff --git a/test/index.js b/test/index.js
index aed1a11..75ab625 100644
--- a/test/index.js
+++ b/test/index.js
@@ -13,8 +13,7 @@ scenario.runTape("description of example test", async (t, { alice }) => {
   // Make a call to a Zome function
   // indicating the function, and passing it an input
   const addr = alice.call("my_zome", "create_my_entry", { "entry": { "content": "sample content" } })
-  const result = alice.call("my_zome", "get_my_entry", { "address": addr.Ok })
 
   // check for equality of the actual and expected results
-  t.deepEqual(result, { Ok: { App: ['my_entry', '{"content":"sample content"}'] } })
+  t.deepEqual(addr, { Ok: "QmXyrzTd876ohgqVtXAn5oLCLAhYoGijoEifH2uQHrCsZ9" })
 })
diff --git a/zomes/my_zome/code/src/lib.rs b/zomes/my_zome/code/src/lib.rs
index 2705ec3..9bc54a3 100644
--- a/zomes/my_zome/code/src/lib.rs
+++ b/zomes/my_zome/code/src/lib.rs
@@ -51,6 +51,7 @@ fn definition() -> ValidatingEntryType {
         },
 
         validation: | _validation_data: hdk::EntryValidationData<MyEntry>| {
+            *hdk::AGENT_ADDRESS;
             Ok(())
         }
     )
```
3. Run `hc test`, it should pass, but it fails because of the dereference at `lib.rs:54`.

**Comment by Qubeo** [comment]
I am also getting (probably) the same error when dereferencing those variables within the "receive" callback.

**Comment by pdaoust** [comment]
@pjkundert is getting the same problems in validation functions for HoloFuel. I see three issues here:

* The `receive()` callback should probably be able to access _all_ statics.
* Validation callbacks should be able to access invariant statics and API functions (e.g., `hdk::DNA_ADDRESS` and `hdk::PROPERTIES`) but not ones that could cause non-deterministic validation behaviour (e.g., `hdk::AGENT_ADDRESS`, which is different for the entry creator and all validators, and `hdk::get_entry()`, which is non-deterministic).
* Attempts to access 'unsafe' things in validation callbacks should get a helpful explainy error.

**Comment by pdaoust** [comment]
@pjkundert 's parallel issue is being tracked in #1445.

**Comment by pdaoust** [comment]
@MightyAlex200 @Qubeo I've got good news -- fixes for two issues are in the works.

* Better error messaging for `BadCallError` will land in 0.0.18 with PR #1443 
* Fix for being unable to access globals in #1450 (not part of a release yet and I don't know how extensively it's been tested)

Keep in mind that, even after #1450 is merged, it's still not a wise idea to use things that vary per agent (like `hdk::AGENT_ADDRESS`) in validation functions, and eventually it'll probably throw an error.

**Comment by willemolding** [comment]
Both #1443 and #1450 have been merged so closing
